futures-util/src/stream/copy_into.rs

```diff
@@ -0,0 +1,81 @@
+use futures_core::future::Future;
+use futures_core::task::{Context, Poll};
+use futures_core::Stream;
+use futures_io::AsyncWrite;
+use std::io;
+use std::pin::Pin;
+
+/// Future for the [`copy_into`](super::StreamExt::copy_into) method.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct CopyInto<'a, S: Unpin, W: ?Sized + Unpin> {
+    stream: S,
+    read_done: bool,
+    writer: &'a mut W,
+    pos: usize,
+    cap: usize,
+    amt: u64,
+    buf: Option<Vec<u8>>,
+}
+
+impl<S: Unpin, W: ?Sized + Unpin> Unpin for CopyInto<'_, S, W> {}
+
+impl<'a, S: Unpin, W: ?Sized + Unpin> CopyInto<'a, S, W> {
+    pub(super) fn new(stream: S, writer: &'a mut W) -> Self {
+        CopyInto {
+            stream,
+            read_done: false,
```

Comment:
I'd keep an `Option<Stream>` rather than a separate boolean, which will allow dropping the stream sooner as well as using less space if the stream type happens to have a free niche.